File: GFG/DP/46.Longest-Bitonic-subsequence.cpp
// Longest Bitonic subsequence

class Solution{
	public:
	
	int LongestBitonicSequence(vector<int>nums)
	{
	    // code here
	    int n = nums.size();
	    vector<int> dp1(n,1);
	    for(int i=0; i<n; i++) {
	        for(int prev=0; prev<=i-1; prev++) {
	            if(nums[i]>nums[prev] && 1+dp1[prev]>dp1[i]) {
	                dp1[i] = 1 + dp1[prev];
	            }
	        }
	    }
	    
	    
	    vector<int> dp2(n,1);
	    for(int i=n-1; i>=0; i--) {
	        for(int prev=n-1; prev>=i+1; prev--) {
	            if(nums[i]>nums[prev] && 1+dp2[prev]>dp2[i]) {
	                dp2[i] = 1 + dp2[prev];
	            }
	        }
	    }
	    
	    int ans = 1;
	    for(int i=0; i<n; i++) {
	        ans = max(ans,dp1[i]+dp2[i]-1);
	    }
	    return ans;
	}
};

File: GFG/DP/46.Longest-Bitonic-subsequence.java
// Longest Bitonic subsequence

class Solution {
    public static int longestBitonicSequence(int n, int[] nums) {
        int[] lis = new int[n];
        int[] lds = new int[n];
        Arrays.fill(lis, 1);
        Arrays.fill(lds, 1);

        // LIS
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < i; j++) {
                if (nums[j] < nums[i]) {
                    lis[i] = Math.max(lis[i], 1 + lis[j]);
                }
            }
        }

        // LDS
        for (int i = n - 1; i >= 0; i--) {
            for (int j = n - 1; j > i; j--) {
                if (nums[j] < nums[i]) {
                    lds[i] = Math.max(lds[i], 1 + lds[j]);
                }
            }
        }

        int ans = 0;
        for (int i = 0; i < n; i++) {
            // both increasing and decreasing parts exist
            if (lis[i] > 1 && lds[i] > 1) {
                ans = Math.max(ans, lis[i] + lds[i] - 1);
            }
        }

        return ans;
    }
}
